pkg/store/bucket.go

```diff
@@ -2210,40 +2211,35 @@ func (r *bucketIndexReader) ExpandedPostings(ctx context.Context, ms []*labels.M
 		return postings, nil
 	}
 	var (
-		postingGroups []*postingGroup
-		allRequested  = false
-		hasAdds       = false
-		keys          []labels.Label
+		allRequested = false
+		hasAdds      = false
+		keys         []labels.Label
 	)
 
-	// NOTE: Derived from tsdb.PostingsForMatchers.
-	for _, m := range ms {
-		if ctx.Err() != nil {
-			return nil, ctx.Err()
-		}
-		// Each group is separate to tell later what postings are intersecting with what.
-		pg, err := toPostingGroup(ctx, r.block.indexHeaderReader.LabelValues, m)
-		if err != nil {
-			return nil, errors.Wrap(err, "toPostingGroup")
-		}
-
-		// If this groups adds nothing, it's an empty group. We can shortcut this, since intersection with empty
-		// postings would return no postings anyway.
-		// E.g. label="non-existing-value" returns empty group.
-		if !pg.addAll && len(pg.addKeys) == 0 {
-			r.storeExpandedPostingsToCache(ms, index.EmptyPostings(), 0)
-			return nil, nil
-		}
-
-		postingGroups = append(postingGroups, pg)
+	postingGroups, err := matchersToPostingGroups(ctx, r.block.indexHeaderReader.LabelValues, ms)
+	if err != nil {
+		return nil, errors.Wrap(err, "matchersToPostingGroups")
+	}
+	if postingGroups == nil {
+		r.storeExpandedPostingsToCache(ms, index.EmptyPostings(), 0)
+		return nil, nil
+	}
+	for _, pg := range postingGroups {
 		allRequested = allRequested || pg.addAll
 		hasAdds = hasAdds || len(pg.addKeys) > 0
 
+		// Sort keys to make sure good performance when fetching posting offsets.
+		slices.Sort(pg.addKeys)
+		slices.Sort(pg.removeKeys)
```

Comment:
Question: Should we try to return these keys sorted from the "merge" method (i think is possible to return them sorted as both inputs already have them sorted?). I'm asking this as the removeKeys can be quite large with regex expressions?

Reply:
I don't sort input keys anymore because later I use a set to deduplicate. Since a posting group might merge multiple times, returned sorted keys from `merge` might be inefficient...
If I don't use set based deduplication then maybe it is better to keep them always sorted

Reply:
Discussed with @alanprot, I will try to not use map for merge/dedup and make sure `merge` returns sorted results

Reply:
This should be fixed. PTAL